toxic_collection.go

```diff
@@ -43,58 +38,137 @@ func (c *ToxicCollection) ResetToxics() {
 	c.Lock()
 	defer c.Unlock()
 
-	for index, toxic := range c.toxics {
-		toxic.SetEnabled(false)
-		c.setToxic(toxic, index)
+	for _, toxic := range c.toxics {
+		// TODO do this in bulk
+		c.chainRemoveToxic(toxic)
 	}
+	c.toxics = c.toxics[:0]
 }
 
-func (c *ToxicCollection) GetToxicMap() map[string]Toxic {
-	result := make(map[string]Toxic)
+func (c *ToxicCollection) GetToxic(name string) toxics.Toxic {
+	c.Lock()
+	defer c.Unlock()
+
 	for _, toxic := range c.toxics {
-		result[toxic.Name()] = toxic
+		if toxic.Name == name {
+			return toxic
+		}
+	}
+	return nil
+}
+
+func (c *ToxicCollection) GetToxicMap() map[string]toxics.Toxic {
+	c.Lock()
+	defer c.Unlock()
+
+	result := make(map[string]toxics.Toxic)
+	for _, toxic := range c.toxics {
+		result[toxic.Name] = toxic
 	}
 	return result
 }
 
-func (c *ToxicCollection) SetToxicJson(name string, data io.Reader) (Toxic, error) {
+func (c *ToxicCollection) AddToxicJson(data io.Reader) (toxics.Toxic, error) {
```

Comment:
It's in here because it needs the lock on ToxicCollection (and because it was in here before...)
The json decoding is a little messy since it needs to figure out what type of toxic it is before it can do the decode (requires access to ToxicCollection)